src/BenchmarkDotNet.TestAdapter/BenchmarkExecutor.cs

```diff
@@ -0,0 +1,86 @@
+        /// Runs all the benchmarks in the given source, updating the TestExecutionRecorder as they get run.
+        /// </summary>
+        /// <param name="source">The dll or exe of the benchmark project.</param>
+        /// <param name="recorder">The interface used to record the current test execution progress.</param>
+        /// <param name="benchmarkIds">
+        /// An optional list of benchmark IDs specifying which benchmarks to run.
+        /// These IDs are the same as the ones generated for the VSTest TestCase.
+        /// </param>
+        public void RunBenchmarks(string source, TestExecutionRecorderWrapper recorder, HashSet<Guid>? benchmarkIds = null)
+        {
+            var benchmarks = BenchmarkEnumerator.GetBenchmarksFromSource(source);
+            var testCases = new List<TestCase>();
+
+            var filteredBenchmarks = new List<BenchmarkRunInfo>();
+            foreach (var benchmark in benchmarks)
+            {
+                var needsJobInfo = benchmark.BenchmarksCases.Select(c => c.Job.DisplayInfo).Distinct().Count() > 1;
+                var filteredCases = new List<BenchmarkCase>();
+                foreach (var benchmarkCase in benchmark.BenchmarksCases)
+                {
+                    var testId = benchmarkCase.GetTestCaseId();
+                    if (benchmarkIds != null && benchmarkIds.Contains(testId))
+                    {
+                        filteredCases.Add(benchmarkCase);
+                        testCases.Add(benchmarkCase.ToVSTestCase(source, needsJobInfo));
+                    }
+                }
+
+                if (filteredCases.Count > 0)
+                {
+                    filteredBenchmarks.Add(new BenchmarkRunInfo(filteredCases.ToArray(), benchmark.Type, benchmark.Config));
+                }
+            }
+
+            benchmarks = filteredBenchmarks.ToArray();
+
+            if (benchmarks.Length == 0)
+                return;
+
+            // Create an event processor which will subscribe to events and push them to VSTest
+            var eventProcessor = new VSTestEventProcessor(testCases, recorder, cts.Token);
+
+            // Create a logger which will forward all log messages in BDN to the VSTest logger.
+            var logger = new VSTestLogger(recorder.GetLogger());
+
+            // Modify all the benchmarks so that the event process and logger is added.
+            benchmarks = benchmarks
+                .Select(b => new BenchmarkRunInfo(
+                    b.BenchmarksCases,
+                    b.Type,
+                    b.Config.AddEventProcessor(eventProcessor).AddLogger(logger).CreateImmutableConfig()))
+                .ToArray();
+
+            // Run all the benchmarks, and ensure that any tests that don't have a result yet are sent.
+            BenchmarkRunner.Run(benchmarks);
+            eventProcessor.SendUnsentTestResults();
+        }
+
+        /// <summary>
+        /// Stop the benchmarks when next able.
```

Comment:
It's possible to stop the run? WOW that is impressive! 🥇

Reply:
Yep! But it does seem a bit hacky abusing the EventHandler to throw an exception to cause this rather than proper cancellation support. Perhaps this would be good to add proper support for in the future.